_episodes_rmd/05-testthat.Rmd

```diff
@@ -0,0 +1,182 @@
+to a known return value (outputs).
+In the function comments, we have already formalised a few of these as examples.
+Copy-and-paste them into the test file, remove the `# should return [1]`
+indicators of the expected results, and wrap the result values into a `c()` to
+enable R to compare the expected values with the output of the two `center()` tests.
+
+```{r center-test, eval=FALSE}
+test_that("centering works", {
+  expect_equal(center(c(1, 2, 3)), c(-1, 0, 1))
+  expect_equal(center(c(1, 2, 3), 1), c(0, 1, 2))
+})
+```
+
+Don't worry that there is no output when you execute either one or both of the
+`expect_equal()` statements, or  the whole `test_that()` block. In good UNIX
+tradition, `testthat` does not bother you with confirming that things are working
+as expected. But there will be plenty of output when some expectations are not
+met, i.e. one or more unit tests fail.
+
+> ## Testing our rescaling functions
+>
+> Apply the above example of converting the examples into unit tests for the
+> `rescale` function as well.
+>
+> > ## Solution
+> > ~~~
+> > test_that("rescaling works", {
+> >   expect_equal(rescale(c(1, 2, 3)), c(0.0, 0.5, 1.0))
+> >   expect_equal(rescale(c(1, 2, 3, 4, 5)), c(0.0, 0.25, 0.5, 0.75, 1.0))
+> > })
+> > ~~~
+> > {: .r}
+> {: .solution}
+{: .challenge}
+
+To conclude this section about creating unit tests, let's again commit our results,
+for example as "Span safety net for TDD".
+
+Finally, have a look at the websites listed on `help(usethis)` to learn about
+how else it can assist you when building packages.
+
+
+## Test-driven development (TDD)
+
+Remember that we updated `rescale()` with lower and upper bounds and default
+values at the end of [episode 2][02]? We had to manually test that change with a
+new example back then. We were repeating ourselves a bit more often than necessary
+back then, weren't we? Of course, there are ways to automate the testing of code
+changes, and to give you quick feedback whether your changes worked, or broke
+anything.
+
+To give ourselves this quick feedback, use `testthat`'s `auto_test_package()`
+and notice the hopefully all green and `OK` `Results`.
+
+With this safety net enabled, we will first update `test-rescale.R`, and then
+update the code in `rescale.R` This strategy of (re)writing (new)
+tests before (re)writing the code-to-be-tested is called
+"[test-driven design/development][TDD]". It is intended to reduce confirmation
+bias when coding. If one already worked hard to get the code to run, one may be
+_biased to_ let the tests _confirm_ what the code currently _does_, instead of
+challenging it with tests about what it _should_ do. The latter is similar to the
+scientific method of trying to disprove hypotheses, so that the truth remains.
+
+[02]: {{ page.root }}/03-func-R/
+[TDD]: https://en.wikipedia.org/wiki/Test-driven_development
+
+> ## Update `test-rescale.R` with the second example from the roxygen comments at the end of
+[episode 2][02].
```

Comment:
Will this render properly if on a new line? I don't think it will.

Reply:
Fix on another branch in a larger link-update commit.